backend-rust/src/graphql_api.rs

```diff
@@ -743,13 +747,189 @@ LIMIT 30", // WHERE slot_time > (LOCALTIMESTAMP - $1::interval)
+
+        let snapshot = ContractSnapshot {
+            block_height: row.block_height,
+            contract_address_index,
+            contract_address_sub_index,
+            contract_name: row.contract_name,
+            module_reference: row.module_reference,
+            amount: row.amount,
+        };
+
+        Ok(Contract {
+            contract_address_index,
+            contract_address_sub_index,
+            contract_address: format!(
+                "<{},{}>",
+                contract_address_index, contract_address_sub_index
+            ),
+            creator: row.creator.into(),
+            block_height: row.block_height,
+            transaction_hash: row.transaction_hash,
+            block_slot_time: row.block_slot_time,
+            snapshot,
+        })
+    }
+
+    async fn contracts<'a>(
+        &self,
+        ctx: &Context<'a>,
+        #[graphql(desc = "Returns the first _n_ elements from the list.")] first: Option<i64>,
+        #[graphql(desc = "Returns the elements in the list that come after the specified cursor.")]
+        after: Option<String>,
+        #[graphql(desc = "Returns the last _n_ elements from the list.")] last: Option<i64>,
+        #[graphql(desc = "Returns the elements in the list that come before the specified cursor.")]
+        before: Option<String>,
+    ) -> ApiResult<connection::Connection<String, Contract>> {
+        let config = get_config(ctx)?;
+        let pool = get_pool(ctx)?;
+        let query = ConnectionQuery::<i64>::new(
+            first,
+            after,
+            last,
+            before,
+            config.contract_connection_limit,
+        )?;
+
+        let mut row_stream = sqlx::query!(
+            r#"
+SELECT * FROM (
+    SELECT
+        contracts.index as index,
+        sub_index,
+        module_reference,
+        name as contract_name,
+        contracts.amount,
+        blocks.slot_time as block_slot_time,
+        init_block_height as block_height,
+        transactions.hash as transaction_hash,
+        accounts.address as creator
+    FROM contracts
+    JOIN blocks ON init_block_height=blocks.height
+    JOIN transactions ON init_block_height=transactions.block_height AND init_transaction_index=transactions.index
+    JOIN accounts ON transactions.sender=accounts.index
+  WHERE contracts.index > $1 AND contracts.index < $2
+  ORDER BY
+     (CASE WHEN $4 THEN contracts.index END) DESC,
+     (CASE WHEN NOT $4 THEN contracts.index END) ASC
+  LIMIT $3
+) AS contract_data
+ORDER BY contract_data.index ASC"#,
+            query.from,
+            query.to,
+            query.limit,
+            query.desc
+        )
+        .fetch(pool);
+
+        let mut connection = connection::Connection::new(true, true);
+
+        while let Some(row) = row_stream.try_next().await? {
+            let contract_address_index = row.index.try_into()?;
+            let contract_address_sub_index = row.sub_index.try_into()?;
```

Comment:
I agree that there is something seriously wrong if our database includes `invalid` data, but returning in error in that case is always better (even it should never happen) than using `unwrap`.

Reply:
I think crashing the API service is a bit aggressive, but it might be worth catching these internal errors in a middleware of the service, and log them as errors.
I'll create a separate task for this.